### PR TITLE
Prevent infinitely recursive wrapper script

### PR DESCRIPTION
**Pull request by ThunderKey**
## Problem

I have `./bin` instead of `bin` in my `PATH` variable. This is not expected by the wrapper script and it lead to the script calling itself all the time.

## Changes

1. At first check if the `$CRYSTAL` command is the current file. If so try to remove `bin` and the script root from the `PATH` and call itself again. If the `PATH` did not change: print an error and abort. This prevents the same check to be necessary when loading the installed library path.
2. Compare the `realpath` with the script path. This prevents having to checking for `bin/crystal` as well and handles `./bin/crystal`
3. Use `$CRYSTAL` instead of `crystal` for loading the installed library path to not create an infinite recursion there

If there is interest in doing so I could also remove `"./bin"` from the path, but it is not necessary since there is a workaround using the `CRYSTAL` env variable.

### CODE REVIEW

**Review by straight-shoota — COMMENTED** [review]
Could you add the explanatory comments into the script directly? It would help understand the function of the code without having to look up this PR.

**Comment by ThunderKey** [comment]
@straight-shoota That's a good idea. I've added some comments. Hopefully this makes it more clear

**Review by straight-shoota — APPROVED** [review]
Thanks :+1:

File: bin/crystal [hunk 1]
```diff
@@ -146,11 +146,22 @@ export CRYSTAL_HAS_WRAPPER=true
...
+if [ "$(realpath "$(command -v "$CRYSTAL")")" = "$SCRIPT_PATH" ]; then
+  # remove the path to this script from PATH
+  NEW_PATH="$(remove_path_item "$(remove_path_item "$PATH" "$SCRIPT_ROOT")" "bin")"
+  # if the PATH did not change it will lead to an infinite recursion => display error
+  if [ "$NEW_PATH" = "$PATH" ]; then
+    __error_msg 'Could not remove the script bin/crystal from the PATH. Remove it by hand or set the CRYSTAL env variable'
+    exit 1
+  fi
+  # re-execute this script with the cleaned up PATH
+  export PATH="$(NEW_PATH)"
```
Comment (beta-ziliani) [inline bin/crystal:159]:
```suggestion
  export PATH="$NEW_PATH"
```

**Review by beta-ziliani — COMMENTED** [review]
There's a variable being wrongly treated as a command

File: bin/crystal (hunk 1 above)
Comment (ThunderKey) [reply bin/crystal]: Oh no, how did I miss that... Thanks for the heads up!

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
This broke the build workflow: https://app.circleci.com/pipelines/github/crystal-lang/crystal/11025/workflows/9ca69536-1cad-49d4-aed1-64b8a80196df/jobs/73844

**Comment by straight-shoota** [comment]
So, the root of the problem is that https://github.com/crystal-lang/shards/blob/master/Makefile calls the wrapper script as defined in the `CRYSTAL` variable, but it also *exports* that variable. So it ends up calling the wrapper script like this: `CRYSTAL=bin/crystal bin/crystal`. That's of course bad and should not be expected to work. But apparently, before this change it did. 🤷
I suppose this needs to be fixed in the `Makefile`, though. It should not export `$CRYSTAL` when calling the wrapper script.

**Comment by HertzDevil** [comment]
It seems #12959 removes that export

**Comment by straight-shoota** [comment]
Eh, no that's a different export. I'm talking about the Makefile for shards exporting `CRYSTAL` variable.

**Comment by straight-shoota** [comment]
It looks like this and #13032 break some valid use cases, for example in https://github.com/crystal-lang/test-ecosystem where no global `crystal` is installed and `PATH` points to `bin/` to allow calling the wrapper script as `crystal`. That should be totally fine as long as `.build/crystal` exists. The only limitation is that the wrapper script must not try to call the parent crystal command when it points to itself.

**Pull request merged**